Codechange: Use std::vector for GRFConfig lists.

PeterN opened the pull request:
## Motivation / Problem

GRFConfig lists are stored as C-style linked lists.

~~And often it is difficult to know if a function accepts a pointer to a list (GRFConfig *) or a pointer to an individual config (GRFConfig *).~~

## Description

Instead, use a `std::vector` of `std::unique_ptr`s for `GRFConfigList`.

~~There is still some faffing about with iterators to work out 'index' for GUI purposes, but some manual management is now gone.~~

~~Some conversions from linked-list to std algorithms are missing or not optimal, as it takes some effort to work out what the mix of pointers and pointers to pointers is actually intended to do.~~

~~To do:~~

* [x] ~~Standardize on parameter naming, differentiating between lists and individual configs.~~
* [x] ~~Fix initial loading which still uses new/delete and makes a copy, instead of doing it in-place.~~
* [x] ~~Fix doxygen which probably refers to pointers and heads and tails.~~

## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/gfxinit.cpp:
```diff
@@ -215,9 +214,8 @@ static void LoadSpriteTables()
-	extra->next = top;
-	master->next = extra;
-	_grfconfig = master;
+	_grfconfig.insert(_grfconfig.begin(), *extra);
```
What about something like `GRFConfig *master = _grfconfig.emplace(_grfconfig.begin(), master_filename);` around like 197?
Saves a copy here and a manual delete later on.
And if/when tackling `GRFConfig *extra`, the `.c_str()` isn't needed anymore since `GRFConfig` accepts `std::string`.

rubidium42 commented on src/network/core/network_game_info.cpp:
```diff
@@ -230,17 +230,16 @@ void SerializeNetworkGameInfo(Packet &p, const NetworkServerGameInfo &info, bool
-		for (c = info.grfconfig; c != nullptr; c = c->next) {
+		for (const auto &c : info.grfconfig) {
 			if (!HasBit(c->flags, GCF_STATIC)) count++;
 		}
```
`uint count = std::ranges::count_if(info.grfconfig, [](const auto &c){ return !HasBit(c->flags, GCF_STATIC); });`

rubidium42 commented on src/newgrf_config.cpp:
```diff
@@ -537,51 +515,24 @@ bool GRFFileScanner::AddFile(const std::string &filename, size_t basepath_length
-		if (_all_grfs == nullptr) {
-			_all_grfs = c;
-		} else {
-			/* Insert file into list at a position determined by its
-			 * name, so the list is sorted as we go along */
-			GRFConfig **pd, *d;
-			bool stop = false;
-			for (pd = &_all_grfs; (d = *pd) != nullptr; pd = &d->next) {
-				if (c->ident.grfid == d->ident.grfid && c->ident.md5sum == d->ident.md5sum) added = false;
-				/* Because there can be multiple grfs with the same name, make sure we checked all grfs with the same name,
-				 *  before inserting the entry. So insert a new grf at the end of all grfs with the same name, instead of
-				 *  just after the first with the same name. Avoids doubles in the list. */
-				if (StrCompareIgnoreCase(c->GetName(), d->GetName()) <= 0) {
-					stop = true;
-				} else if (stop) {
-					break;
-				}
```
Just for the next time reviewing: sorting is *also* done by `DoScanNewGRFFiles`, making all attempts to sort here kinda pointless.

rubidium42 commented on src/newgrf_config.cpp:
```diff
@@ -537,51 +515,24 @@ bool GRFFileScanner::AddFile(const std::string &filename, size_t basepath_length
-				*pd = c;
-			}
+		if (std::ranges::none_of(_all_grfs, [&c](const auto &gc) { return c->ident.grfid == gc->ident.grfid && c->ident.md5sum == gc->ident.md5sum; })) {
+			grfconfig = _all_grfs.emplace_back(std::move(c)).get();
```
This isn't needed. The result of `.get()` on the unique_ptr will not change when moving. Otherwise putting unique_ptrs into vectors would be an enormous hassle.

rubidium42 commented on src/saveload/newgrf_sl.cpp:
```diff
@@ -91,11 +91,11 @@ struct NGRFChunkHandler : ChunkHandler {
 			this->SaveParameters(*c);
 			SlSetArrayIndex(index++);
-			SlObject(c, description);
+			SlObject(&*c, description);
```
Maybe, so it's consistent with `LoadCommon`.
```suggestion
			SlObject(c.get(), description);
```

rubidium42 commented on src/script/api/script_newgrf.cpp:
```diff
@@ -28,11 +28,8 @@ ScriptNewGRFList::ScriptNewGRFList()
+	auto it = std::ranges::find_if(_grfconfig, [grfid](const auto &c) { return !HasBit(c->flags, GCF_STATIC) && c->ident.grfid == grfid; });
+	if (it != std::end(_grfconfig)) return true;
 
 	return false;
```
Maybe use `std::ranges_any_of` instead?

rubidium42 commented on src/settings.cpp:
```diff
@@ -1123,23 +1122,15 @@ static GRFConfig *GRFLoadConfig(const IniFile &ini, const char *grpname, bool is
-		}
-		if (duplicate) {
-			delete c;
+		auto found = std::ranges::find_if(list, [&c](const auto &gc) { return gc->ident.grfid == c->ident.grfid; });
+		if (found != std::end(list)) {
```
```suggestion
		auto found = std::ranges::any_of(list, [&c](const auto &gc) { return gc->ident.grfid == c->ident.grfid; });
		if (found) {
```

PeterN replied:
Can't for this one as `found` is used inside the if-block.

PeterN committed:
Codechange: Use std::vector for GRFConfig lists.

This replaces the C-style custom managed linked-list and allows use of iterators etc.

rubidium42 commented on src/newgrf_gui.cpp:
```diff
@@ -1520,30 +1485,26 @@ struct NewGRFWindow : public Window, NewGRFScanCallback {
-		/* Insert GRF config to configuration list. */
-		c->next = *entry;
-		*entry = c;
+		/*/ Copy GRF details from scanned list. */
```
```suggestion
		/* Copy GRF details from scanned list. */
```
